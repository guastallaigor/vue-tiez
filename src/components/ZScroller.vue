<template>
  <div class="z-scroller" :class="{ dark }" :style="width ? `width:${width}` : ''">
    <div class="inner"><slot/></div>
  </div>
</template>

<script>
export default {
  name: 'z-scroller',
  props: {
    dark: {
      type: Boolean,
      default: false
    },
    width: {
      type: String,
      default: 'auto'
    }
  },
  data: () => ({})
}
</script>

<style lang="scss">
@mixin set-pseudo-element($bgImage: linear-gradient(to right, rgb(255, 255, 255), rgba(255, 255, 255, 0)), $isRight: true) {
  content: "";
  position: absolute;
  top: 0;
  height: 100%;
  width: 20px;
  background-image: $bgImage;

  @if ($isRight == true) {
    right: 0;
  }

  @else {
    left: 0;
  }
}

.z-scroller {
  position: relative;

  &::before {
    @include set-pseudo-element(linear-gradient(to right, rgb(255, 255, 255), rgba(255, 255, 255, 0)), false);
  }

  &::after {
    @include set-pseudo-element();
  }

  .inner {
    display: flex;
    overflow-x: auto;

    section {
      scroll-snap-align: center;
      height: 200px;
      width: 150px;
      display: flex;
      align-items: center;
      justify-content: center;
      background: rgb(239, 239, 239);
      margin: 0 20px;
    }
  }

  &.dark {
    .inner {
      section {
        background: #333;
        color: #fff;
      }
    }

    &::before {
      @include set-pseudo-element(linear-gradient(to right, rgba(0, 0, 0, 0.3), rgba(255, 255, 255, 0)), false);
    }

    &::after {
      @include set-pseudo-element(linear-gradient(to left, rgba(0, 0, 0, 0.3), rgba(255, 255, 255, 0)));
    }
  }
}
</style>
