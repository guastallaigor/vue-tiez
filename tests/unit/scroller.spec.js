import { shallowMount } from '@vue/test-utils'
import ZScroller from '../../src/components/ZScroller'

describe('When I create the ZScroller component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZScroller, {
      propsData,
      slots: {
        default: slot || 'ZScroller'
      }
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent()
    expect(wrapper.find(ZScroller).isVueInstance()).toBe(true)
  })

  it('should have a width auto and dark false by default', () => {
    const wrapper = createComponent()
    const scroller = wrapper.find('.z-scroller')
    expect(scroller.exists()).toBe(true)
    expect(scroller.classes().length).toBe(1)
    expect(scroller.classes()).toEqual(['z-scroller'])
    expect(scroller.attributes().style).toBe('width: auto;')
  })

  it('should change the width to 300px', () => {
    const wrapper = createComponent({ width: '300px' })
    const scroller = wrapper.find('.z-scroller')
    expect(scroller.exists()).toBe(true)
    expect(scroller.classes().length).toBe(1)
    expect(scroller.attributes().style).toBe('width: 300px;')
  })

  it('should add a dark class', () => {
    const wrapper = createComponent({ dark: true })
    const scroller = wrapper.find('.z-scroller')
    expect(scroller.exists()).toBe(true)
    expect(scroller.classes().length).toBe(2)
    expect(scroller.classes()).toEqual(['z-scroller', 'dark'])
  })

  it('should have a slot', () => {
    const wrapper = createComponent({}, '<span>test with slot</span>')
    const scroller = wrapper.find('.z-scroller')
    expect(scroller.exists()).toBe(true)
    expect(scroller.classes().length).toBe(1)
    expect(scroller.classes()).toEqual(['z-scroller'])
    expect(scroller.html()).toBe('<div class="z-scroller" style="width: auto;"><div class="inner"><span>test with slot</span></div></div>')
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      width: 'auto',
      dark: false
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
