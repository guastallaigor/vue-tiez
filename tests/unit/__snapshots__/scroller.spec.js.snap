// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZScroller component should match snapshot 1`] = `
<div class="z-scroller" style="width: auto;">
  <div class="inner">ZScroller</div>
</div>
`;
